internal/watcher/instance_watcher_service.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) F5, Inc.
+//
+// This source code is licensed under the Apache License, Version 2.0 license found in the
+// LICENSE file in the root directory of this source tree.
+
+package watcher
+
+import (
+	"context"
+	"log/slog"
+	"time"
+
+	"github.com/nginx/agent/v3/api/grpc/mpi/v1"
+	"github.com/nginx/agent/v3/internal/config"
+	"github.com/nginx/agent/v3/internal/logger"
+	"github.com/nginx/agent/v3/internal/model"
+)
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6@v6.8.1 -generate
+//counterfeiter:generate . processOperator
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6@v6.8.1 -generate
+//counterfeiter:generate . processParser
+
+type (
+	processOperator interface {
+		Processes(ctx context.Context) ([]*model.Process, error)
+	}
+
+	processParser interface {
+		Parse(ctx context.Context, processes []*model.Process) []*v1.Instance
+	}
+
+	InstanceWatcherService struct {
+		agentConfig     *config.Config
+		processOperator processOperator
+		processParsers  []processParser
+		cache           []*v1.Instance
+	}
+
+	InstanceUpdates struct {
+		newInstances     []*v1.Instance
+		deletedInstances []*v1.Instance
+	}
+)
+
+func NewInstanceWatcherService(agentConfig *config.Config) *InstanceWatcherService {
+	return &InstanceWatcherService{
+		agentConfig:     agentConfig,
+		processOperator: NewProcessOperator(),
+		processParsers: []processParser{
+			NewNginxProcessParser(),
+		},
+		cache: []*v1.Instance{},
+	}
+}
+
+func (iw *InstanceWatcherService) Watch(ctx context.Context, ch chan<- InstanceUpdates) {
+	monitoringFrequency := iw.agentConfig.Watchers.InstanceWatcher.MonitoringFrequency
+	slog.DebugContext(ctx, "Starting instance watcher monitoring", "monitoring_frequency", monitoringFrequency)
+
+	instanceWatcherTicker := time.NewTicker(monitoringFrequency)
+	defer instanceWatcherTicker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			close(ch)
+			return
+		case <-instanceWatcherTicker.C:
+			newCtx := context.WithValue(ctx, logger.CorrelationIDContextKey, logger.GenerateCorrelationID())
+			instanceUpdates, err := iw.updates(newCtx)
+			if err != nil {
+				slog.ErrorContext(newCtx, "Instance watcher updates", "error", err)
+			} else {
+				ch <- instanceUpdates
+			}
+		}
+	}
+}
+
+func (iw *InstanceWatcherService) updates(ctx context.Context) (
+	instanceUpdates InstanceUpdates,
+	err error,
+) {
+	processes, err := iw.processOperator.Processes(ctx)
+	if err != nil {
+		return instanceUpdates, err
+	}
+
+	instancesFound := []*v1.Instance{}
```

Comment:
Is there a way where we could save the PID and/or the PPID of NGINX processes in a map and then check against those before we parse into instances? In other words, we wouldn't perform the Process -> Instance conversion unless there are new instances and we could potentially remove instances without even needing to parse into instances.

You could also only parse the processes that are new instead of the whole processes list.

All of this may just be premature optimization though.

Reply:
Its not going to help much I'd say. At this stage we dont know which processes are NGINX or not. So we need to call `Parse` to filter out the NGINX processes and get the process info. From what I have seen before the parsing of the NGINX process is fast. The slow piece is getting the full process list.

Reply:
Fair enough, let's leave it as is 👍